File: fons/norma-ts/hal/solum.ts
/**
 * solum.ts - File System Implementation
 *
 * Native TypeScript implementation of the HAL solum (filesystem) interface.
 * Uses Bun file APIs and Node.js fs/path/os modules.
 *
 * Verb conjugation encodes sync/async:
 *   - Imperative (-a, -e, -i): synchronous
 *   - Future indicative (-et, -ebit): asynchronous (returns Promise)
 */

import * as fs from 'node:fs/promises';
import * as fsSync from 'node:fs';
import * as path from 'node:path';
import * as os from 'node:os';

/** Full file status returned by describe/describet */
export interface SolumStatus {
    modus: number;        // permission bits (e.g., 0o755)
    nexus: number;        // hard link count
    possessor: number;    // owner uid
    grex: number;         // group gid
    magnitudo: number;    // size in bytes
    modificatum: number;  // mtime (ms since epoch)
    estDirectorii: boolean;
    estVinculum: boolean; // is symlink
}

export const solum = {
    // =========================================================================
    // READING - Text
    // =========================================================================
    // Verb: lege/leget from "legere" (to read, gather)

    /** Read entire file as text (sync) */
    lege(via: string): string {
        return fsSync.readFileSync(via, 'utf-8');
    },

    /** Read entire file as text (async) */
    async leget(via: string): Promise<string> {
        return Bun.file(via).text();
    },

    // =========================================================================
    // READING - Bytes
    // =========================================================================
    // Verb: hauri/hauriet from "haurire" (to draw up, draw water)

    /** Draw entire file as bytes (sync) */
    hauri(via: string): Uint8Array {
        return new Uint8Array(fsSync.readFileSync(via));
    },

    /** Draw entire file as bytes (async) */
    async hauriet(via: string): Promise<Uint8Array> {
        return new Uint8Array(await Bun.file(via).arrayBuffer());
    },

    // =========================================================================
    // READING - Lines
    // =========================================================================
    // Verb: carpe/carpiet from "carpere" (to pluck, pick, harvest)

    /** Pluck lines from file (sync) */
    carpe(via: string): string[] {
        const content = fsSync.readFileSync(via, 'utf-8');
        return content.split(/\r?\n/);
    },

    /** Pluck lines from file (async) */
    async carpiet(via: string): Promise<string[]> {
        const content = await Bun.file(via).text();
        return content.split(/\r?\n/);
    },

    // =========================================================================
    // WRITING - Text
    // =========================================================================
    // Verb: scribe/scribet from "scribere" (to write)

    /** Write text to file, overwrites existing (sync) */
    scribe(via: string, data: string): void {
        fsSync.writeFileSync(via, data, 'utf-8');
    },

    /** Write text to file, overwrites existing (async) */
    async scribet(via: string, data: string): Promise<void> {
        await Bun.write(via, data);
    },

    // =========================================================================
    // WRITING - Bytes
    // =========================================================================
    // Verb: funde/fundet from "fundere" (to pour, pour out)

    /** Pour bytes to file, overwrites existing (sync) */
    funde(via: string, data: Uint8Array): void {
        fsSync.writeFileSync(via, data);
    },

    /** Pour bytes to file, overwrites existing (async) */
    async fundet(via: string, data: Uint8Array): Promise<void> {
        await Bun.write(via, data);
    },

    // =========================================================================
    // WRITING - Append
    // =========================================================================
    // Verb: appone/apponet from "apponere" (to place near, add to)

    /** Append text to file (sync) */
    appone(via: string, data: string): void {
        fsSync.appendFileSync(via, data);
    },

    /** Append text to file (async) */
    async apponet(via: string, data: string): Promise<void> {
        await fs.appendFile(via, data);
    },

    // =========================================================================
    // FILE INFO - Existence
    // =========================================================================
    // Verb: exstat/exstabit from "exstare" (to stand out, exist)

    /** Check if path exists (sync) */
    exstat(via: string): boolean {
        try {
            fsSync.accessSync(via);
            return true;
        }
        catch {
            return false;
        }
    },

    /** Check if path exists (async) */
    async exstabit(via: string): Promise<boolean> {
        try {
            await fs.access(via);
            return true;
        }
        catch {
            return false;
        }
    },

    // =========================================================================
    // FILE INFO - Details
    // =========================================================================
    // Verb: describe/describet from "describere" (to describe, delineate)

    /** Get file details (sync) */
    describe(via: string): SolumStatus {
        const stats = fsSync.lstatSync(via);
        return {
            modus: stats.mode & 0o7777,
            nexus: stats.nlink,
            possessor: stats.uid,
            grex: stats.gid,
            magnitudo: stats.size,
            modificatum: stats.mtimeMs,
            estDirectorii: stats.isDirectory(),
            estVinculum: stats.isSymbolicLink(),
        };
    },

    /** Get file details (async) */
    async describet(via: string): Promise<SolumStatus> {
        const stats = await fs.lstat(via);
        return {
            modus: stats.mode & 0o7777,
            nexus: stats.nlink,
            possessor: stats.uid,
            grex: stats.gid,
            magnitudo: stats.size,
            modificatum: stats.mtimeMs,
            estDirectorii: stats.isDirectory(),
            estVinculum: stats.isSymbolicLink(),
        };
    },

    // =========================================================================
    // FILE INFO - Symlinks
    // =========================================================================
    // Verb: sequere/sequetur from "sequi" (to follow)

    /** Follow symlink to get target path (sync) */
    sequere(via: string): string {
        return fsSync.readlinkSync(via);
    },

    /** Follow symlink to get target path (async) */
    async sequetur(via: string): Promise<string> {
        return fs.readlink(via);
    },

    // =========================================================================
    // FILE OPERATIONS - Delete
    // =========================================================================
    // Verb: dele/delet from "delere" (to destroy, delete)

    /** Delete file (sync) */
    dele(via: string): void {
        fsSync.unlinkSync(via);
    },

    /** Delete file (async) */
    async delet(via: string): Promise<void> {
        await fs.unlink(via);
    },

    // =========================================================================
    // FILE OPERATIONS - Copy
    // =========================================================================
    // Verb: exscribe/exscribet from "exscribere" (to copy out, transcribe)

    /** Copy file (sync) */
    exscribe(fons: string, destinatio: string): void {
        fsSync.copyFileSync(fons, destinatio);
    },

    /** Copy file (async) */
    async exscribet(fons: string, destinatio: string): Promise<void> {
        await fs.copyFile(fons, destinatio);
    },

    // =========================================================================
    // FILE OPERATIONS - Rename/Move
    // =========================================================================
    // Verb: renomina/renominabit from "renominare" (to rename)

    /** Rename or move file (sync) */
    renomina(fons: string, destinatio: string): void {
        fsSync.renameSync(fons, destinatio);
    },

    /** Rename or move file (async) */
    async renominabit(fons: string, destinatio: string): Promise<void> {
        await fs.rename(fons, destinatio);
    },

    // =========================================================================
    // FILE OPERATIONS - Touch
    // =========================================================================
    // Verb: tange/tanget from "tangere" (to touch)

    /** Touch file - create or update mtime (sync) */
    tange(via: string): void {
        const now = new Date();
        try {
            fsSync.utimesSync(via, now, now);
        }
        catch {
            fsSync.writeFileSync(via, '');
        }
    },

    /** Touch file - create or update mtime (async) */
    async tanget(via: string): Promise<void> {
        const now = new Date();
        try {
            await fs.utimes(via, now, now);
        }
        catch {
            await Bun.write(via, '');
        }
    },

    // =========================================================================
    // DIRECTORY OPERATIONS - Create
    // =========================================================================
    // Verb: crea/creabit from "creare" (to create, bring forth)

    /** Create directory, recursive (sync) */
    crea(via: string): void {
        fsSync.mkdirSync(via, { recursive: true });
    },

    /** Create directory, recursive (async) */
    async creabit(via: string): Promise<void> {
        await fs.mkdir(via, { recursive: true });
    },

    // =========================================================================
    // DIRECTORY OPERATIONS - List
    // =========================================================================
    // Verb: enumera/enumerabit from "enumerare" (to count out, enumerate)

    /** List directory contents (sync) */
    enumera(via: string): string[] {
        return fsSync.readdirSync(via);
    },

    /** List directory contents (async) */
    async enumerabit(via: string): Promise<string[]> {
        return fs.readdir(via);
    },

    // =========================================================================
    // DIRECTORY OPERATIONS - Prune/Remove
    // =========================================================================
    // Verb: amputa/amputabit from "amputare" (to cut off, prune)

    /** Prune directory tree, recursive (sync) */
    amputa(via: string): void {
        fsSync.rmSync(via, { recursive: true, force: true });
    },

    /** Prune directory tree, recursive (async) */
    async amputabit(via: string): Promise<void> {
        await fs.rm(via, { recursive: true, force: true });
    },

    // =========================================================================
    // PATH UTILITIES
    // =========================================================================
    // Pure functions on path strings, not filesystem I/O. Sync only.

    /** Join path segments */
    iunge(partes: string[]): string {
        return path.join(...partes);
    },

    /** Get directory part of path */
    directorium(via: string): string {
        return path.dirname(via);
    },

    /** Get filename part of path */
    basis(via: string): string {
        return path.basename(via);
    },

    /** Get file extension (includes dot) */
    extensio(via: string): string {
        return path.extname(via);
    },

    /** Resolve to absolute path */
    absolve(via: string): string {
        return path.resolve(via);
    },

    /** Get user's home directory */
    domus(): string {
        return os.homedir();
    },

    /** Get system temp directory */
    temporarium(): string {
        return os.tmpdir();
    },
};

File: fons/norma-py/hal/solum.py
"""solum.py - File System Implementation

Native Python implementation of the HAL solum (filesystem) interface.
Uses os and pathlib for sync, aiofiles for async.

Verb conjugation encodes sync/async:
  - Imperative (-a, -e, -i): synchronous
  - Future indicative (-et, -ebit): asynchronous (returns awaitable)
"""

import os
import asyncio
import shutil
import tempfile
from dataclasses import dataclass
from pathlib import Path
from typing import Optional

try:
    import aiofiles
    import aiofiles.os
    _HAS_AIOFILES = True
except ImportError:
    _HAS_AIOFILES = False


@dataclass
class SolumStatus:
    """Full file status returned by describe/describet."""
    modus: int           # permission bits (e.g., 0o755)
    nexus: int           # hard link count
    possessor: int       # owner uid
    grex: int            # group gid
    magnitudo: int       # size in bytes
    modificatum: int     # mtime (ms since epoch)
    est_directorii: bool
    est_vinculum: bool   # is symlink


# =============================================================================
# READING - Text
# =============================================================================
# Verb: lege/leget from "legere" (to read, gather)

def lege(via: str) -> str:
    """Read entire file as text (sync)."""
    return Path(via).read_text()


async def leget(via: str) -> str:
    """Read entire file as text (async)."""
    if _HAS_AIOFILES:
        async with aiofiles.open(via, "r") as f:
            return await f.read()
    loop = asyncio.get_event_loop()
    return await loop.run_in_executor(None, lege, via)


# =============================================================================
# READING - Bytes
# =============================================================================
# Verb: hauri/hauriet from "haurire" (to draw up, draw water)

def hauri(via: str) -> bytes:
    """Draw entire file as bytes (sync)."""
    return Path(via).read_bytes()


async def hauriet(via: str) -> bytes:
    """Draw entire file as bytes (async)."""
    if _HAS_AIOFILES:
        async with aiofiles.open(via, "rb") as f:
            return await f.read()
    loop = asyncio.get_event_loop()
    return await loop.run_in_executor(None, hauri, via)


# =============================================================================
# READING - Lines
# =============================================================================
# Verb: carpe/carpiet from "carpere" (to pluck, pick, harvest)

def carpe(via: str) -> list[str]:
    """Pluck lines from file (sync)."""
    return Path(via).read_text().splitlines()


async def carpiet(via: str) -> list[str]:
    """Pluck lines from file (async)."""
    content = await leget(via)
    return content.splitlines()


# =============================================================================
# WRITING - Text
# =============================================================================
# Verb: scribe/scribet from "scribere" (to write)

def scribe(via: str, data: str) -> None:
    """Write text to file, overwrites existing (sync)."""
    Path(via).write_text(data)


async def scribet(via: str, data: str) -> None:
    """Write text to file, overwrites existing (async)."""
    if _HAS_AIOFILES:
        async with aiofiles.open(via, "w") as f:
            await f.write(data)
        return
    loop = asyncio.get_event_loop()
    await loop.run_in_executor(None, scribe, via, data)


# =============================================================================
# WRITING - Bytes
# =============================================================================
# Verb: funde/fundet from "fundere" (to pour, pour out)

def funde(via: str, data: bytes) -> None:
    """Pour bytes to file, overwrites existing (sync)."""
    Path(via).write_bytes(data)


async def fundet(via: str, data: bytes) -> None:
    """Pour bytes to file, overwrites existing (async)."""
    if _HAS_AIOFILES:
        async with aiofiles.open(via, "wb") as f:
            await f.write(data)
        return
    loop = asyncio.get_event_loop()
    await loop.run_in_executor(None, funde, via, data)


# =============================================================================
# WRITING - Append
# =============================================================================
# Verb: appone/apponet from "apponere" (to place near, add to)

def appone(via: str, data: str) -> None:
    """Append text to file (sync)."""
    with open(via, "a") as f:
        f.write(data)


async def apponet(via: str, data: str) -> None:
    """Append text to file (async)."""
    if _HAS_AIOFILES:
        async with aiofiles.open(via, "a") as f:
            await f.write(data)
        return
    loop = asyncio.get_event_loop()
    await loop.run_in_executor(None, appone, via, data)


# =============================================================================
# FILE INFO - Existence
# =============================================================================
# Verb: exstat/exstabit from "exstare" (to stand out, exist)

def exstat(via: str) -> bool:
    """Check if path exists (sync)."""
    return Path(via).exists()


async def exstabit(via: str) -> bool:
    """Check if path exists (async)."""
    if _HAS_AIOFILES:
        return await aiofiles.os.path.exists(via)
    loop = asyncio.get_event_loop()
    return await loop.run_in_executor(None, exstat, via)


# =============================================================================
# FILE INFO - Details
# =============================================================================
# Verb: describe/describet from "describere" (to describe, delineate)

def describe(via: str) -> SolumStatus:
    """Get file details (sync)."""
    stat = os.lstat(via)
    return SolumStatus(
        modus=stat.st_mode & 0o7777,
        nexus=stat.st_nlink,
        possessor=stat.st_uid,
        grex=stat.st_gid,
        magnitudo=stat.st_size,
        modificatum=int(stat.st_mtime * 1000),
        est_directorii=os.path.isdir(via),
        est_vinculum=os.path.islink(via),
    )


async def describet(via: str) -> SolumStatus:
    """Get file details (async)."""
    loop = asyncio.get_event_loop()
    return await loop.run_in_executor(None, describe, via)


# =============================================================================
# FILE INFO - Symlinks
# =============================================================================
# Verb: sequere/sequetur from "sequi" (to follow)

def sequere(via: str) -> str:
    """Follow symlink to get target path (sync)."""
    return os.readlink(via)


async def sequetur(via: str) -> str:
    """Follow symlink to get target path (async)."""
    loop = asyncio.get_event_loop()
    return await loop.run_in_executor(None, sequere, via)


# =============================================================================
# FILE OPERATIONS - Delete
# =============================================================================
# Verb: dele/delet from "delere" (to destroy, delete)

def dele(via: str) -> None:
    """Delete file (sync)."""
    os.remove(via)


async def delet(via: str) -> None:
    """Delete file (async)."""
    if _HAS_AIOFILES:
        await aiofiles.os.remove(via)
        return
    loop = asyncio.get_event_loop()
    await loop.run_in_executor(None, dele, via)


# =============================================================================
# FILE OPERATIONS - Copy
# =============================================================================
# Verb: exscribe/exscribet from "exscribere" (to copy out, transcribe)

def exscribe(fons: str, destinatio: str) -> None:
    """Copy file (sync)."""
    shutil.copy2(fons, destinatio)


async def exscribet(fons: str, destinatio: str) -> None:
    """Copy file (async)."""
    loop = asyncio.get_event_loop()
    await loop.run_in_executor(None, exscribe, fons, destinatio)


# =============================================================================
# FILE OPERATIONS - Rename/Move
# =============================================================================
# Verb: renomina/renominabit from "renominare" (to rename)

def renomina(fons: str, destinatio: str) -> None:
    """Rename or move file (sync)."""
    os.rename(fons, destinatio)


async def renominabit(fons: str, destinatio: str) -> None:
    """Rename or move file (async)."""
    if _HAS_AIOFILES:
        await aiofiles.os.rename(fons, destinatio)
        return
    loop = asyncio.get_event_loop()
    await loop.run_in_executor(None, renomina, fons, destinatio)


# =============================================================================
# FILE OPERATIONS - Touch
# =============================================================================
# Verb: tange/tanget from "tangere" (to touch)

def tange(via: str) -> None:
    """Touch file - create or update mtime (sync)."""
    Path(via).touch()


async def tanget(via: str) -> None:
    """Touch file - create or update mtime (async)."""
    loop = asyncio.get_event_loop()
    await loop.run_in_executor(None, tange, via)


# =============================================================================
# DIRECTORY OPERATIONS - Create
# =============================================================================
# Verb: crea/creabit from "creare" (to create, bring forth)

def crea(via: str) -> None:
    """Create directory, recursive (sync)."""
    Path(via).mkdir(parents=True, exist_ok=True)


async def creabit(via: str) -> None:
    """Create directory, recursive (async)."""
    if _HAS_AIOFILES:
        await aiofiles.os.makedirs(via, exist_ok=True)
        return
    loop = asyncio.get_event_loop()
    await loop.run_in_executor(None, crea, via)


# =============================================================================
# DIRECTORY OPERATIONS - List
# =============================================================================
# Verb: enumera/enumerabit from "enumerare" (to count out, enumerate)

def enumera(via: str) -> list[str]:
    """List directory contents (sync)."""
    return os.listdir(via)


async def enumerabit(via: str) -> list[str]:
    """List directory contents (async)."""
    if _HAS_AIOFILES:
        return await aiofiles.os.listdir(via)
    loop = asyncio.get_event_loop()
    return await loop.run_in_executor(None, enumera, via)


# =============================================================================
# DIRECTORY OPERATIONS - Prune/Remove
# =============================================================================
# Verb: amputa/amputabit from "amputare" (to cut off, prune)

def amputa(via: str) -> None:
    """Prune directory tree, recursive (sync)."""
    shutil.rmtree(via, ignore_errors=True)


async def amputabit(via: str) -> None:
    """Prune directory tree, recursive (async)."""
    loop = asyncio.get_event_loop()
    await loop.run_in_executor(None, amputa, via)


# =============================================================================
# PATH UTILITIES
# =============================================================================
# Pure functions on path strings, not filesystem I/O. Sync only.

def iunge(partes: list[str]) -> str:
    """Join path segments."""
    return os.path.join(*partes) if partes else ""


def directorium(via: str) -> str:
    """Get directory part of path."""
    return os.path.dirname(via)


def basis(via: str) -> str:
    """Get filename part of path."""
    return os.path.basename(via)


def extensio(via: str) -> str:
    """Get file extension (includes dot)."""
    _, ext = os.path.splitext(via)
    return ext


def absolve(via: str) -> str:
    """Resolve to absolute path."""
    return os.path.abspath(via)


def domus() -> str:
    """Get user's home directory."""
    return os.path.expanduser("~")


def temporarium() -> str:
    """Get system temp directory."""
    return tempfile.gettempdir()
